Fixing the shallow copy problem in JPetEvent

wkrzemien opened the pull request:
-Add clone() methods to JPetXXXHit hierarchy according to clone idiom (see e.g. https://stackoverflow.com/questions/12255546/c-deep-copying-a-base-class-pointer) 
-add copy constructor, assignment operator to JPetEvent
-change member container for hits from raw to unique pointers.
-refactor other methods to rather copy the hits instead of taking the pointers directly.
-change the definition of JPetAnalysisTools function that orders the hits in time. 

The code is not fully tested yet, but I wanted to ask for your opinion @alekgajos @kkacprzak

wkrzemien committed:
Add clone methods to all Hit classes

The clone methods are necessary to implement virtual clone pattern
that will be used in JPetEvent class to correctly copy hits.

wkrzemien committed:
Change part of JPetEvent methods to copy hits

Also, add missing clone methods to JPetPhysHit

wkrzemien committed:
Add copy constructor,= operator to JPetEvent

Also, change the methods to copy hits and refactor JPetAnalysisTools
method.

alekgajos reviewed (commented):
Thank you! This definitely solves the problem, bit I am just wondering if we could simplify the access to the hits' vector in JPetEvent, especially avoid having `getHits2()` (or we should at least give it a more descriptive name).

As far as I know, we never modify the vector of hits inside an event in user code. What do you think about skipping direct access fo JPetEvent::getHIts entirely?

wkrzemien committed:
Add second version of getHitsOrderedByTime

First version operates with bare pointers and returns a new ordered
vector. The second version operates on unique_ptr and orders
vector in-place. Also, add tests for second version.

wkrzemien committed:
Remove methods getHits2 and getHitsCopy

They are not needed.
